File: src/main/java/jugglinglab/core/PatternListPanel.java
// PatternListPanel.java
//
// Copyright 2002-2023 Jack Boyce and the Juggling Lab contributors

package jugglinglab.core;

import java.awt.*;
import java.awt.datatransfer.*;
import java.awt.event.*;
import java.util.ArrayList;
import java.util.ResourceBundle;
import javax.swing.*;
import javax.swing.border.BevelBorder;
import javax.swing.event.*;

import jugglinglab.jml.*;
import jugglinglab.jml.JMLPatternList.PatternRecord;
import jugglinglab.util.*;
import jugglinglab.view.View;


public class PatternListPanel extends JPanel {
    static final ResourceBundle guistrings = jugglinglab.JugglingLab.guistrings;
    static final ResourceBundle errorstrings = jugglinglab.JugglingLab.errorstrings;

    static final Font font_nopattern = new Font("SanSerif", Font.BOLD | Font.ITALIC, 14);
    static final Font font_pattern = new Font("Monospaced", Font.PLAIN, 14);
    static final Font font_pattern_popup = new Font("Monospaced", Font.ITALIC, 14);

    static final DataFlavor patternFlavor = new DataFlavor(PatternRecord.class,
                                                 "Juggling Lab pattern record");

    protected JMLPatternList pl;
    protected JFrame parent;
    protected View animtarget;
    protected JList<PatternRecord> list;

    // for mouse/popup menu handling
    protected boolean didPopup;
    protected ArrayList<PatternWindow> popupPatterns;
    protected JDialog dialog;
    protected JTextField tf;
    protected JButton okbutton;

    // for drag and drop operations
    protected boolean draggingOut;


    protected PatternListPanel() {
        pl = new JMLPatternList();
        makePanel();
        setOpaque(false);
    }

    public PatternListPanel(JFrame parent) {
        this();
        this.parent = parent;
    }

    public PatternListPanel(View target) {
        this();
        setTargetView(target);
    }

    //-------------------------------------------------------------------------
    // Methods to create and manage contents
    //-------------------------------------------------------------------------

    protected void makePanel() {
        list = new JList<PatternRecord>(pl.getModel());
        list.getSelectionModel().setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
        list.setCellRenderer(new PatternCellRenderer());

        list.setDragEnabled(true);
        list.setTransferHandler(new PatternTransferHandler());

        JScrollPane pane = new JScrollPane(list);

        list.addMouseListener(new MouseAdapter() {
            @Override
            public void mousePressed(final MouseEvent me) {
                int row = list.locationToIndex(me.getPoint());
                if (row >= 0)
                    list.setSelectedIndex(row);

                didPopup = false;

                if (me.isPopupTrigger()) {
                    // On macOS the popup triggers here
                    makePopupMenu().show(list, me.getX(), me.getY());
                    didPopup = true;
                }
            }

            @Override
            public void mouseReleased(final MouseEvent me) {
                if (me.isPopupTrigger()) {
                    // On Windows the popup triggers here
                    makePopupMenu().show(list, me.getX(), me.getY());
                    didPopup = true;
                }

                if (!didPopup) {
                    launchAnimation();
                    checkSelection();
                }
            }
        });

        setLayout(new BorderLayout());
        add(pane, BorderLayout.CENTER);
    }

    // Try to launch an animation window for the currently-selected item in the
    // list. If there is no pattern associated with the line, do nothing.
    protected void launchAnimation() {
        try {
            int row = list.getSelectedIndex();
            if (row < 0 || (JMLPatternList.BLANK_AT_END && row == pl.getModel().size() - 1))
                return;

            JMLPattern pat = pl.getPatternForLine(row);
            if (pat == null)
                return;
            pat.layoutPattern();  // do this before getting hash code
            if (PatternWindow.bringToFront(pat.getHashCode()))
                return;

            AnimationPrefs ap = pl.getAnimationPrefsForLine(row);

            if (animtarget != null)
                animtarget.restartView(pat, ap);
            else
                new PatternWindow(pat.getTitle(), pat, ap);
        } catch (JuggleExceptionUser jeu) {
            new ErrorDialog(PatternListPanel.this, jeu.getMessage());
        } catch (JuggleExceptionInternal jei) {
            ErrorDialog.handleFatalException(jei);
        }
    }

    public JMLPatternList getPatternList() {
        return pl;
    }

    public void clearList() {
        pl.clearModel();
    }

    public void setTargetView(View target) {
        animtarget = target;
    }

    // Used by GeneratorTarget
    public void addPattern(String display, String animprefs, String notation, String anim) {
        pl.addLine(-1, display, animprefs, notation, anim, null, null);
    }

    //-------------------------------------------------------------------------
    // Popup menu and associated handler methods
    //-------------------------------------------------------------------------

    protected static final String[] popupItems =
        {
            "PLPOPUP Insert text...",
            null,
            "PLPOPUP Insert pattern",
            null,
            "PLPOPUP Change title...",
            "PLPOPUP Change display text...",
            null,
            "PLPOPUP Remove line",
        };

    protected static final String[] popupCommands =
        {
            "inserttext",
            null,
            "insertpattern",
            null,
            "title",
            "displaytext",
            null,
            "remove",
        };

    protected JPopupMenu makePopupMenu() {
        JPopupMenu popup = new JPopupMenu();
        int row = list.getSelectedIndex();

        popupPatterns = new ArrayList<PatternWindow>();
        for (Frame fr : Frame.getFrames()) {
            if (fr.isVisible() && fr instanceof PatternWindow)
                popupPatterns.add((PatternWindow)fr);
        }

        ActionListener al = new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ae) {
                String command = ae.getActionCommand();
                int row = list.getSelectedIndex();

                if (command.equals("inserttext")) {
                    insertText(row);
                } else if (command.equals("insertpattern")) {
                    // do nothing
                } else if (command.equals("title")) {
                    changeTitle();
                } else if (command.equals("displaytext")) {
                    changeDisplayText(row);
                } else if (command.equals("remove")) {
                    pl.getModel().remove(row);
                } else {
                    // inserting a pattern
                    int patnum = Integer.parseInt(command.substring(3, command.length()));
                    PatternWindow pw = popupPatterns.get(patnum);
                    insertPattern(row, pw);
                }
            }
        };

        for (int i = 0; i < popupItems.length; ++i) {
            String name = popupItems[i];

            if (name == null) {
                popup.addSeparator();
                continue;
            }

            JMenuItem item = new JMenuItem(guistrings.getString(name.replace(' ', '_')));
            item.setActionCommand(popupCommands[i]);
            item.addActionListener(al);

            if ((popupCommands[i].equals("displaytext") || popupCommands[i].equals("remove")) && row < 0)
                item.setEnabled(false);
            if ((popupCommands[i].equals("displaytext") || popupCommands[i].equals("remove")) &&
                        JMLPatternList.BLANK_AT_END && row == pl.getModel().size() - 1)
                item.setEnabled(false);
            if (popupCommands[i].equals("insertpattern") && popupPatterns.size() == 0)
                item.setEnabled(false);

            popup.add(item);

            if (popupCommands[i].equals("insertpattern")) {
                int patnum = 0;
                for (PatternWindow pw : popupPatterns) {
                    JMenuItem pitem = new JMenuItem("   " + pw.getTitle());
                    pitem.setActionCommand("pat" + patnum);
                    pitem.addActionListener(al);
                    pitem.setFont(font_pattern_popup);
                    popup.add(pitem);
                    ++patnum;
                }
            }
        }

        popup.setBorder(new BevelBorder(BevelBorder.RAISED));

        popup.addPopupMenuListener(new PopupMenuListener() {
            @Override
            public void popupMenuCanceled(PopupMenuEvent e) { checkSelection(); }

            @Override
            public void popupMenuWillBecomeInvisible(PopupMenuEvent e) {}

            @Override
            public void popupMenuWillBecomeVisible(PopupMenuEvent e) {}
        });

        return popup;
    }

    // Insert the pattern in the given PatternWindow into the given row
    protected void insertPattern(int row, PatternWindow pw) {
        String display = pw.getTitle();
        String animprefs = pw.getAnimationPrefs().toString();
        if (animprefs.length() == 0)
            animprefs = null;

        String notation = "jml";
        String anim = null;

        JMLPattern pattern = pw.getPattern();
        JMLNode patnode = null;
        try {
            patnode = pattern.getRootNode().findNode("pattern");
        } catch (JuggleExceptionInternal jei) {
            // any error here cannot be user error since pattern is
            // already animating in another window
            ErrorDialog.handleFatalException(jei);
            return;
        }
        JMLNode infonode = patnode.findNode("info");

        if (pattern.hasBasePattern() && !pattern.isBasePatternEdited()) {
            // add as base pattern instead of JML
            notation = pattern.getBasePatternNotation();
            anim = pattern.getBasePatternConfig();
            patnode = null;
        }

        pl.addLine(row, display, animprefs, notation, anim, patnode, infonode);

        if (row < 0) {
            if (JMLPatternList.BLANK_AT_END)
                list.setSelectedIndex(pl.getModel().size() - 2);
            else
                list.setSelectedIndex(pl.getModel().size() - 1);
        } else
            list.setSelectedIndex(row);
    }

    // Open a dialog to allow the user to insert a line of text
    protected void insertText(int row) {
        makeDialog(guistrings.getString("PLDIALOG_Insert_text"), "");

        okbutton.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                String display = tf.getText();
                dialog.dispose();

                pl.addLine(row, display, null, null, null, null, null);

                if (row < 0)
                    list.setSelectedIndex(pl.size() - 1);
                else
                    list.setSelectedIndex(row);
            }
        });

        dialog.setVisible(true);
    }

    // Open a dialog to allow the user to change the pattern list's title
    protected void changeTitle() {
        makeDialog(guistrings.getString("PLDIALOG_Change_title"), pl.getTitle());

        okbutton.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                pl.setTitle(tf.getText());
                dialog.dispose();

                if (parent != null)
                    parent.setTitle(pl.getTitle());
            }
        });

        dialog.setVisible(true);
        list.clearSelection();
    }

    // Open a dialog to allow the user to change the display text of a line
    protected void changeDisplayText(int row) {
        PatternRecord rec = pl.getModel().get(row);
        makeDialog(guistrings.getString("PLDIALOG_Change_display_text"), rec.display);

        okbutton.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                rec.display = tf.getText();
                dialog.dispose();

                pl.getModel().set(row, rec);
            }
        });

        dialog.setVisible(true);
    }

    // Helper to make popup dialog boxes
    protected JDialog makeDialog(String title, String default_text) {
        dialog = new JDialog(parent, title, true);
        GridBagLayout gb = new GridBagLayout();
        dialog.getContentPane().setLayout(gb);

        tf = new JTextField(20);
        tf.setText(default_text);

        okbutton = new JButton(guistrings.getString("OK"));

        dialog.getContentPane().add(tf);
        gb.setConstraints(tf, JLFunc.constraints(GridBagConstraints.LINE_START,0,0,
                                               new Insets(10,10,0,10)));
        dialog.getContentPane().add(okbutton);
        gb.setConstraints(okbutton, JLFunc.constraints(GridBagConstraints.LINE_END,0,1,
                                                     new Insets(10,10,10,10)));
        dialog.getRootPane().setDefaultButton(okbutton);  // OK button is default
        dialog.pack();
        dialog.setResizable(false);
        dialog.setLocationRelativeTo(this);

        dialog.addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) { checkSelection(); }
        });

        return dialog;
    }

    // Be sure to call this at the very end of any mouse-related interaction
    protected void checkSelection() {
        if (JMLPatternList.BLANK_AT_END && list.getSelectedIndex() == pl.getModel().size() - 1)
            list.clearSelection();

        popupPatterns = null;
        dialog = null;
        tf = null;
        okbutton = null;
    }

    //-------------------------------------------------------------------------
    // Classes to support drag and drop operations
    //-------------------------------------------------------------------------

    class PatternTransferHandler extends TransferHandler {
        @Override
        public int getSourceActions(JComponent c) {
            return TransferHandler.COPY_OR_MOVE;
        }

        @Override
        public Transferable createTransferable(JComponent c) {
            int row = list.getSelectedIndex();
            if (row < 0 || (JMLPatternList.BLANK_AT_END && row == pl.getModel().size() - 1))
                return null;

            draggingOut = true;
            PatternRecord rec = pl.getModel().get(row);
            return new PatternTransferable(rec);
        }

        @Override
        public boolean canImport(TransferHandler.TransferSupport info) {
            // support only drop (not clipboard paste)
            if (!info.isDrop())
                return false;

            if (draggingOut)
                info.setDropAction(MOVE);  // within same list
            else
                info.setDropAction(COPY);  // between lists

            if (info.isDataFlavorSupported(patternFlavor))
                return true;
            if (info.isDataFlavorSupported(DataFlavor.stringFlavor))
                return true;

            return false;
        }

        @Override
        public boolean importData(TransferHandler.TransferSupport info) {
            if (!info.isDrop())
                return false;

            JList.DropLocation dl = (JList.DropLocation)info.getDropLocation();
            int index = dl.getIndex();
            if (index < 0)
                index = (JMLPatternList.BLANK_AT_END ? pl.getModel().size() - 1 : pl.getModel().size());

            // Get the record that is being dropped
            Transferable t = info.getTransferable();

            try {
                if (t.isDataFlavorSupported(patternFlavor)) {
                    PatternRecord rec = (PatternRecord)t.getTransferData(patternFlavor);
                    pl.getModel().add(index, new PatternRecord(rec));
                    list.setSelectedIndex(index);
                    return true;
                }

                if (t.isDataFlavorSupported(DataFlavor.stringFlavor)) {
                    String s = (String)t.getTransferData(DataFlavor.stringFlavor);

                    // allow for multi-line strings
                    String[] lines = s.stripTrailing().split("\n");

                    for (int i = lines.length - 1; i >= 0; --i) {
                        PatternRecord rec = new PatternRecord(lines[i],
                                                null, null, null, null, null, null);
                        pl.getModel().add(index, rec);
                    }
                    list.setSelectedIndex(index);
                    return true;
                }
            } catch (Exception e) {
                ErrorDialog.handleFatalException(e);
            }

            return false;
        }

        @Override
        protected void exportDone(JComponent c, Transferable data, int action) {
            if (action == TransferHandler.MOVE) {
                if (!(data instanceof PatternTransferable))
                    return;

                if (!pl.getModel().removeElement(((PatternTransferable)data).rec))
                    ErrorDialog.handleFatalException(
                            new JuggleExceptionInternal("PLP: exportDone()"));
            }

            draggingOut = false;
        }
    }

    class PatternTransferable implements Transferable {
        public PatternRecord rec;

        public PatternTransferable(PatternRecord pr) {
            rec = pr;
        }

        @Override
        public Object getTransferData(DataFlavor flavor) {
            if (flavor.equals(patternFlavor))
                return rec;

            if (flavor.equals(DataFlavor.stringFlavor)) {
                String s;
                if (rec.anim == null || rec.anim.equals(""))
                    s = rec.display;
                else
                    s = rec.anim;
                return s;
            }

            return null;
        }

        @Override
        public DataFlavor[] getTransferDataFlavors() {
            return new DataFlavor[] {
                patternFlavor,
                DataFlavor.stringFlavor,
            };
        }

        @Override
        public boolean isDataFlavorSupported(DataFlavor flavor) {
            return flavor.equals(patternFlavor) || flavor.equals(DataFlavor.stringFlavor);
        }
    }

    //-------------------------------------------------------------------------
    // Class to support rendering of list items
    //-------------------------------------------------------------------------

    class PatternCellRenderer extends JLabel implements ListCellRenderer<PatternRecord> {
        public Component getListCellRendererComponent(
                        JList<? extends PatternRecord> list,  // the list
                        PatternRecord value,     // value to display
                        int index,               // cell index
                        boolean isSelected,      // is the cell selected
                        boolean cellHasFocus)    // does the cell have focus
        {
            PatternRecord rec = value;

            setFont(rec.anim == null && rec.patnode == null ? font_nopattern : font_pattern);
            setText(rec.display.length() > 0 ? rec.display : " ");

            if (isSelected) {
                setBackground(list.getSelectionBackground());
                setForeground(list.getSelectionForeground());
            } else {
                setBackground(list.getBackground());
                setForeground(list.getForeground());
            }
            setEnabled(list.isEnabled());
            setOpaque(true);
            return this;
        }
    }
}

File: src/main/kotlin/jugglinglab/core/PatternListPanel.kt
//
// PatternListPanel.kt
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

@file:Suppress("KotlinConstantConditions", "SimplifyBooleanWithConstants")

package jugglinglab.core

import jugglinglab.JugglingLab.guistrings
import jugglinglab.jml.JMLNode
import jugglinglab.jml.JMLPatternList
import jugglinglab.jml.JMLPatternList.PatternRecord
import jugglinglab.util.ErrorDialog.handleFatalException
import jugglinglab.util.ErrorDialog.handleUserException
import jugglinglab.util.JuggleExceptionInternal
import jugglinglab.util.JuggleExceptionUser
import jugglinglab.util.constraints
import jugglinglab.view.View
import java.awt.*
import java.awt.datatransfer.DataFlavor
import java.awt.datatransfer.Transferable
import java.awt.event.*
import javax.swing.*
import javax.swing.border.BevelBorder
import javax.swing.event.PopupMenuEvent
import javax.swing.event.PopupMenuListener
import kotlin.collections.ArrayList

class PatternListPanel private constructor() : JPanel() {
    val patternList: JMLPatternList = JMLPatternList()

    private var parentFrame: JFrame? = null
    private var animTarget: View? = null
    private lateinit var list: JList<PatternRecord>
    var hasUnsavedChanges: Boolean = false

    // for mouse/popup menu handling
    private var didPopup: Boolean = false
    private var popupPatterns: ArrayList<PatternWindow>? = null
    private var dialog: JDialog? = null
    private var tf: JTextField? = null
    private var okButton: JButton? = null

    // for drag and drop operations
    private var draggingOut: Boolean = false

    init {
        makePanel()
        setOpaque(false)
    }

    constructor(parentFrame: JFrame?) : this() {
        this.parentFrame = parentFrame
    }

    constructor(target: View?) : this() {
        setTargetView(target)
    }

    //--------------------------------------------------------------------------
    // Methods to create and manage contents
    //--------------------------------------------------------------------------

    private fun makePanel() {
        list = JList(patternList.model)
        list.selectionModel.selectionMode = ListSelectionModel.SINGLE_SELECTION
        list.setCellRenderer(PatternCellRenderer())

        list.setDragEnabled(true)
        list.setTransferHandler(PatternTransferHandler())

        val pane = JScrollPane(list)
        list.addMouseListener(
            object : MouseAdapter() {
                override fun mousePressed(me: MouseEvent) {
                    val row = list.locationToIndex(me.getPoint())
                    if (row >= 0) {
                        list.setSelectedIndex(row)
                    }

                    didPopup = false

                    if (me.isPopupTrigger) {
                        // On macOS the popup triggers here
                        makePopupMenu().show(list, me.getX(), me.getY())
                        didPopup = true
                    }
                }

                override fun mouseReleased(me: MouseEvent) {
                    if (me.isPopupTrigger) {
                        // On Windows the popup triggers here
                        makePopupMenu().show(list, me.getX(), me.getY())
                        didPopup = true
                    }

                    if (!didPopup) {
                        launchAnimation()
                        checkSelection()
                    }
                }
            })

        setLayout(BorderLayout())
        add(pane, BorderLayout.CENTER)
    }

    // Try to launch an animation window for the currently-selected item in the
    // list. If there is no pattern associated with the line, do nothing.

    private fun launchAnimation() {
        try {
            val row = list.selectedIndex
            if (row < 0 || (JMLPatternList.BLANK_AT_END && row == patternList.model.size() - 1)) {
                return
            }

            val pat = patternList.getPatternForLine(row) ?: return
            pat.layoutPattern()  // do this before getting hash code
            if (PatternWindow.bringToFront(pat.hashCode)) {
                return
            }

            val ap = patternList.getAnimationPrefsForLine(row)

            if (animTarget != null) {
                animTarget!!.restartView(pat, ap)
            } else {
                PatternWindow(pat.title, pat, ap)
            }
        } catch (jeu: JuggleExceptionUser) {
            handleUserException(this@PatternListPanel, jeu.message)
        } catch (jei: JuggleExceptionInternal) {
            handleFatalException(jei)
        }
    }

    fun clearList() {
        if (patternList.model.size() > (if (JMLPatternList.BLANK_AT_END) 1 else 0)) {
            hasUnsavedChanges = true
        }
        patternList.clearModel()
    }

    fun setTargetView(target: View?) {
        animTarget = target
    }

    // Used by GeneratorTarget.

    fun addPattern(display: String?, animprefs: String?, notation: String?, anim: String?) {
        patternList.addLine(-1, display, animprefs, notation, anim, null, null)
    }

    private fun makePopupMenu(): JPopupMenu {
        val popup = JPopupMenu()
        val row = list.selectedIndex

        val pupatterns = ArrayList<PatternWindow>()
        popupPatterns = pupatterns
        for (fr in Frame.getFrames()) {
            if (fr.isVisible && fr is PatternWindow) {
                pupatterns.add(fr)
            }
        }

        val al =
            ActionListener { ae: ActionEvent? ->
                val command = ae!!.getActionCommand()
                val row1 = list.selectedIndex
                when (command) {
                    "inserttext" -> insertText(row1)
                    "insertpattern" -> {}
                    "displaytext" -> changeDisplayText(row1)
                    "remove" -> patternList.model.remove(row1)
                    else -> {
                        // inserting a pattern
                        val patnum = command.substring(3).toInt()
                        val pw = pupatterns[patnum]
                        insertPattern(row1, pw)
                    }
                }
            }

        for (i in popupItems.indices) {
            val name: String? = popupItems[i]
            if (name == null) {
                popup.addSeparator()
                continue
            }

            val item = JMenuItem(guistrings.getString(name.replace(' ', '_')))
            item.actionCommand = popupCommands[i]
            item.addActionListener(al)

            if ((popupCommands[i] == "displaytext" || popupCommands[i] == "remove") && row < 0) {
                item.setEnabled(false)
            }
            if ((popupCommands[i] == "displaytext" || popupCommands[i] == "remove")
                && JMLPatternList.BLANK_AT_END && row == patternList.model.size() - 1
            ) {
                item.setEnabled(false)
            }
            if (popupCommands[i] == "insertpattern" && pupatterns.isEmpty()) {
                item.setEnabled(false)
            }

            popup.add(item)

            if (popupCommands[i] == "insertpattern") {
                var patnum = 0
                for (pw in pupatterns) {
                    val pitem = JMenuItem("   " + pw.getTitle())
                    pitem.actionCommand = "pat$patnum"
                    pitem.addActionListener(al)
                    pitem.setFont(FONT_PATTERN_POPUP)
                    popup.add(pitem)
                    ++patnum
                }
            }
        }

        popup.setBorder(BevelBorder(BevelBorder.RAISED))

        popup.addPopupMenuListener(
            object : PopupMenuListener {
                override fun popupMenuCanceled(e: PopupMenuEvent?) {
                    checkSelection()
                }
                override fun popupMenuWillBecomeInvisible(e: PopupMenuEvent?) {}
                override fun popupMenuWillBecomeVisible(e: PopupMenuEvent?) {}
            })

        return popup
    }

    // Insert the pattern in the given PatternWindow into the given row.

    private fun insertPattern(row: Int, pw: PatternWindow) {
        val display = pw.getTitle()
        var animprefs: String? = pw.animationPrefs.toString()
        if (animprefs!!.isEmpty()) {
            animprefs = null
        }
        var notation: String? = "jml"
        var anim: String? = null

        val pattern = pw.pattern!!
        var patnode: JMLNode?
        try {
            patnode = pattern.rootNode!!.findNode("pattern")
        } catch (jei: JuggleExceptionInternal) {
            // any error here cannot be user error since pattern is
            // already animating in another window
            handleFatalException(jei)
            return
        }
        val infonode = patnode!!.findNode("info")

        if (pattern.hasBasePattern && !pattern.isBasePatternEdited) {
            // add as base pattern instead of JML
            notation = pattern.basePatternNotation
            anim = pattern.basePatternConfig
            patnode = null
        }

        patternList.addLine(row, display, animprefs, notation, anim, patnode, infonode)

        if (row < 0) {
            if (JMLPatternList.BLANK_AT_END) {
                list.setSelectedIndex(patternList.model.size() - 2)
            } else {
                list.setSelectedIndex(patternList.model.size() - 1)
            }
        } else {
            list.setSelectedIndex(row)
        }

        hasUnsavedChanges = true
    }

    // Open a dialog to allow the user to insert a line of text.

    private fun insertText(row: Int) {
        makeDialog(guistrings.getString("PLDIALOG_Insert_text"), "")

        okButton!!.addActionListener { _: ActionEvent? ->
            val display = tf!!.getText()
            dialog!!.dispose()

            patternList.addLine(row, display, null, null, null, null, null)
            if (row < 0) {
                list.setSelectedIndex(patternList.size - 1)
            } else {
                list.setSelectedIndex(row)
            }
            hasUnsavedChanges = true
        }

        dialog!!.isVisible = true
    }

    // Open a dialog to allow the user to change the display text of a line.

    private fun changeDisplayText(row: Int) {
        val rec = patternList.model.get(row)
        makeDialog(guistrings.getString("PLDIALOG_Change_display_text"), rec.display)

        okButton!!.addActionListener { _: ActionEvent? ->
            rec.display = tf!!.getText()
            dialog!!.dispose()

            patternList.model.set(row, rec)
            hasUnsavedChanges = true
        }

        dialog!!.isVisible = true
    }

    // Helper to make popup dialog boxes.

    private fun makeDialog(title: String?, defaultText: String?): JDialog {
        dialog = JDialog(parentFrame, title, true)
        val d = dialog!!
        val gb = GridBagLayout()
        d.contentPane.setLayout(gb)

        tf = JTextField(20)
        tf!!.text = defaultText

        okButton = JButton(guistrings.getString("OK"))

        d.contentPane.add(tf)
        gb.setConstraints(
            tf, constraints(GridBagConstraints.LINE_START, 0, 0, Insets(10, 10, 0, 10))
        )
        d.contentPane.add(okButton)
        gb.setConstraints(
            okButton,
            constraints(GridBagConstraints.LINE_END, 0, 1, Insets(10, 10, 10, 10))
        )
        d.getRootPane().setDefaultButton(okButton) // OK button is default
        d.pack()
        d.setResizable(false)
        d.setLocationRelativeTo(this)

        d.addWindowListener(
            object : WindowAdapter() {
                override fun windowClosing(e: WindowEvent?) {
                    checkSelection()
                }
            })

        return d
    }

    // Do final cleanup after any mouse-related interaction.

    private fun checkSelection() {
        if (JMLPatternList.BLANK_AT_END && list.selectedIndex == patternList.model.size() - 1) {
            list.clearSelection()
        }
        popupPatterns = null
        dialog = null
        tf = null
        okButton = null
    }

    //--------------------------------------------------------------------------
    // Classes to support drag and drop operations
    //--------------------------------------------------------------------------

    internal inner class PatternTransferHandler : TransferHandler() {
        override fun getSourceActions(c: JComponent?): Int {
            return COPY_OR_MOVE
        }

        override fun createTransferable(c: JComponent?): Transferable? {
            val row = list.selectedIndex
            if (row < 0 || (JMLPatternList.BLANK_AT_END && row == patternList.model.size() - 1)) {
                return null
            }

            draggingOut = true
            val rec = patternList.model.get(row)
            return PatternTransferable(rec)
        }

        override fun canImport(info: TransferSupport): Boolean {
            // support only drop (not clipboard paste)
            if (!info.isDrop) {
                return false
            }

            if (draggingOut) {
                info.setDropAction(MOVE) // within same list
            } else {
                info.setDropAction(COPY) // between lists
            }

            if (info.isDataFlavorSupported(PATTERN_FLAVOR)) {
                return true
            }
            return info.isDataFlavorSupported(DataFlavor.stringFlavor)
        }

        override fun importData(info: TransferSupport): Boolean {
            if (!info.isDrop) {
                return false
            }

            val dl = info.getDropLocation() as JList.DropLocation
            var index = dl.index
            if (index < 0) {
                index = (if (JMLPatternList.BLANK_AT_END) patternList.model.size() - 1 else patternList.model.size())
            }

            // Get the record that is being dropped
            val t = info.getTransferable()

            try {
                if (t.isDataFlavorSupported(PATTERN_FLAVOR)) {
                    val rec = t.getTransferData(PATTERN_FLAVOR) as PatternRecord
                    patternList.model.add(index, PatternRecord(rec))
                    list.setSelectedIndex(index)
                    hasUnsavedChanges = true
                    return true
                }

                if (t.isDataFlavorSupported(DataFlavor.stringFlavor)) {
                    val s = t.getTransferData(DataFlavor.stringFlavor) as String

                    // allow for multi-line strings
                    val lines: Array<String?> =
                        s.trimEnd().split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

                    for (i in lines.indices.reversed()) {
                        val rec = PatternRecord(lines[i]!!, null, null, null, null, null, null)
                        patternList.model.add(index, rec)
                    }
                    list.setSelectedIndex(index)
                    hasUnsavedChanges = true
                    return true
                }
            } catch (e: Exception) {
                handleFatalException(e)
            }

            return false
        }

        override fun exportDone(c: JComponent?, data: Transferable?, action: Int) {
            if (action == MOVE) {
                if (data !is PatternTransferable) {
                    return
                }

                if (!patternList.model.removeElement(data.rec)) {
                    handleFatalException(JuggleExceptionInternal("PLP: exportDone()"))
                }

                hasUnsavedChanges = true
            }

            draggingOut = false
        }
    }

    internal class PatternTransferable(var rec: PatternRecord) : Transferable {
        override fun getTransferData(flavor: DataFlavor): Any {
            if (flavor.equals(PATTERN_FLAVOR)) {
                return rec
            }

            if (flavor.equals(DataFlavor.stringFlavor)) {
                return if (rec.anim == null || rec.anim!!.isEmpty()) {
                    rec.display
                } else {
                    rec.anim!!
                }
            }

            return rec
        }

        override fun getTransferDataFlavors(): Array<DataFlavor?> {
            return arrayOf(
                PATTERN_FLAVOR,
                DataFlavor.stringFlavor,
            )
        }

        override fun isDataFlavorSupported(flavor: DataFlavor): Boolean {
            return flavor.equals(PATTERN_FLAVOR) || flavor.equals(DataFlavor.stringFlavor)
        }
    }

    //--------------------------------------------------------------------------
    // Class to support rendering of list items
    //--------------------------------------------------------------------------

    internal class PatternCellRenderer : JLabel(), ListCellRenderer<PatternRecord?> {
        override fun getListCellRendererComponent(
            list: JList<out PatternRecord?>,  // the list
            rec: PatternRecord?,  // value to display
            index: Int,  // cell index
            isSelected: Boolean,  // is the cell selected
            cellHasFocus: Boolean
        ): Component // does the cell have focus
        {
            if (rec == null) return this
            setFont(if (rec.anim == null && rec.patnode == null) FONT_NOPATTERN else FONT_PATTERN)
            setText(if (!rec.display.isEmpty()) rec.display else " ")

            if (isSelected) {
                setBackground(list.selectionBackground)
                setForeground(list.selectionForeground)
            } else {
                setBackground(list.getBackground())
                setForeground(list.getForeground())
            }
            setEnabled(list.isEnabled)
            setOpaque(true)
            return this
        }
    }

    companion object {
        val FONT_NOPATTERN: Font = Font("SanSerif", Font.BOLD or Font.ITALIC, 14)
        val FONT_PATTERN: Font = Font("Monospaced", Font.PLAIN, 14)
        val FONT_PATTERN_POPUP: Font = Font("Monospaced", Font.ITALIC, 14)

        val PATTERN_FLAVOR: DataFlavor = DataFlavor(PatternRecord::class.java,
            "Juggling Lab pattern record")

        //----------------------------------------------------------------------
        // Popup menu and associated handler methods
        //----------------------------------------------------------------------

        private val popupItems: List<String?> = listOf(
            "PLPOPUP Insert text...",
            null,
            "PLPOPUP Insert pattern",
            null,
            "PLPOPUP Change display text...",
            null,
            "PLPOPUP Remove line",
        )

        private val popupCommands: List<String?> = listOf(
            "inserttext",
            null,
            "insertpattern",
            null,
            "displaytext",
            null,
            "remove",
        )
    }
}
